qmp/libvirtgo_linux.go

```diff
@@ -0,0 +1,327 @@
+	if mon.virConn == nil {
+		return nil, errors.New("Run() need a established connection to proceed.")
+	}
+
+	result, err := qemuMonitorCommandInternal(mon.domain, string(cmd))
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(result), nil
+}
+
+var qemuMonitorCommandInternal = func(
+	domain *libvirt.VirDomain, cmd string) (string, error) {
+	return domain.QemuMonitorCommand(
+		libvirt.VIR_DOMAIN_QEMU_MONITOR_COMMAND_DEFAULT,
+		cmd)
+}
+
+// Events streams QEMU QMP Events.
+// If a problem is encountered setting up the event monitor connection
+// an error will be returned. Errors encountered during streaming will
+// cause the returned event channel to be closed.
+func (mon *libvirtGoMonitorLinux) Events() (<-chan Event, error) {
+
+	if mon.virConn == nil {
+		return nil,
+			errors.New("Events() need a established connection to proceed.")
+	}
+
+	callbackID, err := domainEventRegister(mon)
+	if err != nil {
+		return nil, err
+	}
+
+	go pollEventsLoop(mon.doneChan)
+
+	go domainEventDeregister(mon, callbackID)
+
+	return mon.eventsChan, nil
+}
+
+var lookupDomainByName = func(
+	virConn *libvirt.VirConnection,
+	domainName string) (libvirt.VirDomain, error) {
+	return virConn.LookupDomainByName(domainName)
+}
+
+// eventRegisterDefaultImpl registers a default event implementation.
+func eventRegisterDefaultImpl() {
+	eventRegisterID := eventRegisterDefaultImplInternal()
+	if eventRegisterID == -1 {
+		log.Printf(
+			"EventRegisterDefaultImpl returned an unexpected value %d\n",
+			eventRegisterID)
+		//TODO: panic or what???
+	}
+}
+
+var eventRegisterDefaultImplInternal = func() int {
+	return libvirt.EventRegisterDefaultImpl()
+}
+
+// domainEventRegister register with libvirt to receive events for the
+// specified domain.
+func domainEventRegister(mon *libvirtGoMonitorLinux) (int, error) {
+	callback :=
+		libvirt.DomainEventCallback(newEventCallback(mon))
+
+	//TODO: add ability to register for more event types
+	callbackID := domainEventRegisterInternal(
+		mon,
+		&callback,
+		func() {
+			// empty on purpose until it's decided
+			// what to do here.
+		},
+	)
+	if callbackID == -1 {
+		return -1, errors.New("Domain event registration failed!")
+	}
+
+	return callbackID, nil
+}
+
+var domainEventRegisterInternal = func(
+	mon *libvirtGoMonitorLinux,
+	callback *libvirt.DomainEventCallback,
+	fn func()) int {
+	return mon.virConn.DomainEventRegister(
+		*mon.domain,
+		libvirt.VIR_DOMAIN_EVENT_ID_LIFECYCLE,
+		callback,
+		fn,
+	)
+}
+
+// domainEventDeregister stops the registration with libvirt from receiving
+// events for the specified domain.
+func domainEventDeregister(
+	mon *libvirtGoMonitorLinux,
+	callbackID int) {
+	<-mon.doneChan
+	ret := domainEventDeregisterInternal(mon.virConn, callbackID)
+	if ret != 0 {
+		log.Printf("DomainEventDeregister returned an unexpected value: %d\n",
+			ret)
+	}
+}
+
+var domainEventDeregisterInternal = func(virConn *libvirt.VirConnection,
```

Comment:
Swapping out functions during tests is messy.  Make this an interface and make implementations of that for the monitor and the tests.